app/src/main/java/com/example/habitsmasher/UserValidator.java

```diff
@@ -0,0 +1,155 @@
+    private static final String PASSWORD_LENGTH_MUST_BE_GREATER_THAN_6_MESSAGE = "Password length must be greater than 6!";
+    private static final String PASSWORD_LENGTH_MUST_BE_LESS_THAN_16_MESSAGE = "Password length must be less than 16!";
+    private static final int UPPER_CHARACTER_LIMIT = 15;
+    private static final int LOWER_CHARACTER_LIMIT = 6;
+
+    private EditText _usernameInput;
+    private EditText _emailInput;
+    private EditText _passwordInput;
+
+    String _username;
+    String _email;
+    String _password;
+
+    // for testing
+    public UserValidator() {}
+
+    // for sign up
+    public UserValidator(EditText usernameInput,
+                         EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = usernameInput;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = _usernameInput.getText().toString().trim();
+    }
+
+    // for log in
+    public UserValidator(EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = null;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = "";  // since there is no username for logging in, empty
+    }
+
+    public String getValidUsernameForSignUp() {
+        if (isNewUserValid()) {
+            return _username;
+        }
+        return "";
+    }
+    public String getValidEmailForSignUp() {
+        if (isNewUserValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForSignUp() {
+        if (isNewUserValid()) {
+            return _password;
+        }
+        return "";
+    }
+
+    public String getValidEmailForLogin() {
+        if (isLoginValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForLogin() {
+        if (isLoginValid()) {
+            return _password;
+        }
+        return "";
+    }
+
+    private boolean isNewUserValid() {
+        return isNewUserValid(_username, _email, _password);
+    }
+
+    private boolean isLoginValid() {
+        return isLoginValid(_email, _password);
+    }
+
+    public boolean isNewUserValid(String username, String email, String password) {
+        return !(username.isEmpty() | username.length() > UPPER_CHARACTER_LIMIT | !isLoginValid(email, password));
+    }
+
+    public boolean isLoginValid(String email, String password) {
+        return !(email.isEmpty() | password.isEmpty() |
+                !PatternsCompat.EMAIL_ADDRESS.matcher(email).matches() |
```

Comment:
Just wondering, so this line(103) checks whether an email is valid? If so, does it check for emails that are not just Gmails? For example, would Yahoo and Microsoft emails work? Another thing, would emails that end in different top level domains be accepted as well?

Reply:
Good question! Just ran some tests, and it looks like all email strings are accepted. Yahoo, Microsoft, Outlook, anything. As long as it has an **at sign (@)** and ends in some sort of domain name ending (**..com, .ca, etc.)**